aten/src/ATen/native/quantized/cpu/qupsample_nearest2d.cpp

```diff
@@ -0,0 +1,199 @@
+    for (int64_t h2 = 0; h2 < output_height; ++h2) {
+      const int64_t h1 = h2;
+
+      for (int64_t w2 = 0; w2 < output_width; ++w2) {
+        const int64_t w1 = w2;
+        const auto* pos1 = &i_p[h1 * input_width + w1];
+        auto* pos2 = &o_p[h2 * output_width + w2];
+
+        for (int64_t c = 0; c < channels; ++c) {
+          pos2[0] = pos1[0];
+          pos1 += input_height * input_width;
+          pos2 += output_height * output_width;
+        }
+      }
+    }
+    return;
+  }
+
+  for (int64_t h2 = 0; h2 < output_height; ++h2) {
+    const int64_t h1 =
+        nearest_neighbor_compute_source_index(height_scale, h2, input_height);
+
+    for (int64_t w2 = 0; w2 < output_width; ++w2) {
+      const int64_t w1 =
+          nearest_neighbor_compute_source_index(width_scale, w2, input_width);
+
+      const auto* pos1 = &i_p[h1 * input_width + w1];
+      auto* pos2 = &o_p[h2 * output_width + w2];
+
+      for (int64_t c = 0; c < channels; ++c) {
+        pos2[0] = pos1[0];
+        pos1 += input_height * input_width;
+        pos2 += output_height * output_width;
+      }
+    }
+  }
+}
+
+template <typename scalar_t>
+static void upsample_nearest2d_out_frame_nhwc(
+    scalar_t* odata,
+    scalar_t* idata,
+    int64_t input_height,
+    int64_t input_width,
+    int64_t output_height,
+    int64_t output_width,
+    int64_t nbatch,
+    int64_t channels) {
+  const float height_scale = (float)input_height / (float)output_height;
+  const float width_scale = (float)input_width / (float)output_width;
+  for (int b = 0; b < nbatch; b++) {
+    auto* i_p = reinterpret_cast<typename scalar_t::underlying*>(idata + b * input_height * input_width * channels);
+    auto* o_p = reinterpret_cast<typename scalar_t::underlying*>(odata + b * output_height * output_width * channels);
+    // special case: just copy
+    if (input_height == output_height && input_width == output_width) {
+      for (int64_t h2 = 0; h2 < output_height; ++h2) {
+        const int64_t h1 = h2;
+
+        for (int64_t w2 = 0; w2 < output_width; ++w2) {
+          const int64_t w1 = w2;
+          const auto* pos1 = &i_p[(h1 * input_width + w1) * channels];
+          auto* pos2 = &o_p[(h2 * output_width + w2) * channels];
+
+          for (int64_t c = 0; c < channels; ++c) {
+            pos2[0] = pos1[0];
+            pos1++;
+            pos2++;;
+          }
+        }
+      }
+      return;
+    }
+
+    for (int64_t h2 = 0; h2 < output_height; ++h2) {
+      const int64_t h1 =
+          nearest_neighbor_compute_source_index(height_scale, h2, input_height);
+
+      for (int64_t w2 = 0; w2 < output_width; ++w2) {
+        const int64_t w1 =
+            nearest_neighbor_compute_source_index(width_scale, w2, input_width);
+
+        const auto* pos1 = &i_p[(h1 * input_width + w1)*channels];
+        auto* pos2 = &o_p[(h2 * output_width + w2)*channels];
+
+        for (int64_t c = 0; c < channels; ++c) {
```

Comment:
This is definitely going to be auto-vectorizable. See https://gcc.godbolt.org/z/QTr8AE vs https://gcc.godbolt.org/z/S1ag2z. Maybe we should move this to `kernels/`

Reply:
True, but there is no computation. Just memory access?

Reply:
Using vector instructions makes for cleaner and more efficient memory transfers. For example, see any stdlib's implementation of `memcpy`. It's hand-tuned vector assembly

Reply:
(you can also just call `memcpy` here)